File: 4.depth-first-search/동규/2.diameter-of-binary-tree.js
// https://leetcode.com/problems/diameter-of-binary-tree/
function diameterOfBinaryTree(root) {
  if (!root) return 0;
  let result = -1;
  let getH = function (node, h) {
    if (!node) return 0;
    let left_H = getH(node.left);
    let right_H = getH(node.right);
    result = Math.max(result, 1 + left_H + right_H);
    return 1 + Math.max(left_H, right_H);
  };
  getH(root);
  return result - 1;
}

File: 4.depth-first-search/지연/2.diameter-of-binary-tree.py
# https://leetcode.com/problems/diameter-of-binary-tree/


from typing import Optional

# Definition for a binary tree node.


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.length = 0

        if not root:
            return 0

        def treeLength(root):
            tree_left = treeLength(root.left)
            tree_right = treeLength(root.right)
            self.length = max(self.length, tree_left + tree_right)
            return max(tree_left, tree_right) + 1

        treeLength(root)
        return self.length
